### PR TITLE
Add allowance target test

### PR DESCRIPTION
**Pull request by charlesjhongc**
Add test for `AllowanceTarget`.

### CODE REVIEW

File: contracts/test/AllowanceTarget.t.sol [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    function testSetupAllowance() public {
+        assertEq(allowanceTarget.spender(), address(this));
+    }
+
+    /*********************************
+     *  test: setSpenderWithTimelock *
+     *********************************/
+
+    function testCantSetSpenderWithTimelockByRandomEOA() public {
```
Comment (pilagod) [inline contracts/test/AllowanceTarget.t.sol:43]: nit: test`Cannot`

File: contracts/test/AllowanceTarget.t.sol [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        vm.prank(bob);
+        vm.expectRevert("AllowanceTarget: not the spender");
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+    }
+
+    function testCantSetSpenderWithTimelockWithInvalidAddress() public {
+        vm.expectRevert("AllowanceTarget: new spender not a contract");
+        allowanceTarget.setSpenderWithTimelock(bob);
+        //vm.expectRevert("");
```
Comment (pilagod) [inline contracts/test/AllowanceTarget.t.sol:52]: nit: Could be removed

File: contracts/test/AllowanceTarget.t.sol [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+        bytes memory data;
+        allowanceTarget.executeCall(payable(address(newSpender)), data);
+    }
+
+    function testExecuteCall() public {
+        MockERC20 token = new MockERC20("Test", "TST", 18);
+
+        // mint(address to, uint256 value)
+        allowanceTarget.executeCall(payable(address(token)), abi.encodeWithSelector(MockERC20.mint.selector, address(bob), 1e18));
```
Comment (pilagod) [inline contracts/test/AllowanceTarget.t.sol:132]: Should we check bob balance to ensure allowanceTarget executing the call we expected?

File: contracts/test/AllowanceTarget.t.sol (hunk 1 above)
Comment (NIC619) [reply contracts/test/AllowanceTarget.t.sol]: +1, `Cannot` is clearer than `Cant`

Comment (charlesjhongc) [reply contracts/test/AllowanceTarget.t.sol]: Fixed in 0cd05827df343e0d916a159839ea90389e5fc8b1

File: contracts/test/AllowanceTarget.t.sol [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+     *********************************/
+
+    function testCantTeardownIfNotSpender() public {
+        vm.prank(bob);
+        vm.expectRevert("AllowanceTarget: not the spender");
+        allowanceTarget.teardown();
+    }
+
+    function testTeardown() public {
+        // Seems like the test contract itself have issue with receving ether.
```
Comment (NIC619) [inline contracts/test/AllowanceTarget.t.sol:101]: I think we can add a `receive()` function to test contract

File: contracts/test/AllowanceTarget.t.sol (hunk 2 above)
Comment (charlesjhongc) [reply contracts/test/AllowanceTarget.t.sol]: Fixed in 1e95142359c25c88d73363d06d1f7d15e009673a

File: contracts/test/AllowanceTarget.t.sol [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    function testSetSpenderWithTimelock() public {
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        assertEq(allowanceTarget.newSpender(), newSpender);
+        assertEq(allowanceTarget.timelockExpirationTime(), block.timestamp + 1 days);
+    }
+
+    /*********************************
+     *  test: setSpenderWithTimelock *
```
[the remark this answers is not in the document]
Comment (charlesjhongc) [reply contracts/test/AllowanceTarget.t.sol]: Right, it was missed. Fixed in 1e95142359c25c88d73363d06d1f7d15e009673a

File: contracts/test/AllowanceTarget.t.sol [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        vm.expectRevert("AllowanceTarget: time lock not expired yet");
+        allowanceTarget.completeSetSpender();
+    }
+
+    function testCompleteSetSpender() public {
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        // fast forward
+        vm.warp(allowanceTarget.timelockExpirationTime());
+        allowanceTarget.completeSetSpender();
```
[the remark this answers is not in the document]
Comment (charlesjhongc) [reply contracts/test/AllowanceTarget.t.sol:86]: Make sense. Added in 219124dce1aacf31de59a238fe02e114aaab9e9e

File: contracts/test/AllowanceTarget.t.sol (hunk 3 above)
Comment (charlesjhongc) [reply contracts/test/AllowanceTarget.t.sol]: Added in 219124dce1aacf31de59a238fe02e114aaab9e9e

File: contracts/test/AllowanceTarget.t.sol (hunk 4 above)
Comment (charlesjhongc) [reply contracts/test/AllowanceTarget.t.sol:101]: `selfdestruct` doesn't require `receive` or `fallback` function in fact. I tried adding `receive()` and still not working as expected. Not sure why but mock contract works fine.

Comment (NIC619) [reply contracts/test/AllowanceTarget.t.sol:101]: Got it.

**Review by pilagod — APPROVED** [review]
LGTM 💪

**Pull request merged**